pkg/kube/client/cached.go

```diff
@@ -0,0 +1,256 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	klabels "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pkg/config/schema/kubeclient"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/controllers"
+	"istio.io/istio/pkg/kube/kubetypes"
+	"istio.io/istio/pkg/ptr"
+	"istio.io/pkg/log"
+)
+
+// Cached wraps a Kubernetes client providing cached read access and direct write access.
+// This is based on informers, so most of the same caveats to informers apply here.
+type CachedRead[T controllers.Object] interface {
+	// Get looks up an object by name and namespace. If it does not exist, nil is returned
+	Get(name, namespace string) T
+	// List looks up an object by namespace and labels.
+	// Use metav1.NamespaceAll and klabels.Everything() to select everything.
+	List(namespace string, selector klabels.Selector) []T
+	// ListUnfiltered is like List but ignores any *client side* filters previously configured.
+	ListUnfiltered(namespace string, selector klabels.Selector) []T
+
+	// AddEventHandler inserts a handler. The handler will be called for all Create/Update/Removals.
+	// When ShutdownHandlers is called, the handler is removed.
+	AddEventHandler(h cache.ResourceEventHandler)
+	// HasSynced returns true when the informer is initially populated and that all handlers added
+	// via AddEventHandler have been called with the initial state.
+	// note: this differs from a standard informer HasSynced, which does not check handlers have been called.
+	HasSynced() bool
+	// ShutdownHandlers terminates all handlers added by AddEventHandler.
+	// Warning: this only applies to handlers called via AddEventHandler; any handlers directly added
+	// to the underlying informer are not touched
+	ShutdownHandlers()
+}
+
+type Cached[T controllers.Object] interface {
```

Comment:
WDYT about naming this Wrapped/Wrapper?

Reply:
I don't really like the name wrapped since it's not descriptive and exposes implementation details. Users don't need to know it's wrapping something, they just need to know the functionality. An informer itself is also a wrapper of a bunch of things wrapping others